### PR TITLE
add ability to clone projects

### PR DESCRIPTION
**Pull request by eperedo**
### :pushpin: References

* **Issue:** Closes https://app.clickup.com/t/8695fa0pv

### :memo: Implementation

- [x] Add "Clone" action to table and router.
- [x] Get existing project and reset metadata references (orgUnit, dataSets, documents, created date)
- [x] Get existing documents (blobs) to be re-uploaded in the cloned project
- [x] Clone project

### :art: Screenshots

[Data-Management-App_clone_project.webm](https://github.com/user-attachments/assets/07d9a390-b178-4eb3-b08e-91c65a1a43b8)

#8695fa0pv

### CODE REVIEW

**Comment by ifoche** [comment]
Task linked: [CU-8695fa0pv Ability to clone projects](https://app.clickup.com/t/8695fa0pv)

File: src/pages/project-wizard/ProjectWizard.tsx [hunk 1]
```diff
@@ -220,6 +232,17 @@ class ProjectWizardImpl extends React.Component<Props, State> {
         return await getValidationMessages(this.state.project, currentStep.validationKeys);
     };
 
+    getTitle = (): string => {
+        switch (this.props.action.type) {
+            case "edit":
+                return i18n.t("Edit project");
+            case "clone":
+                return i18n.t("Clone project");
+            default:
```
Comment (tokland) [inline src/pages/project-wizard/ProjectWizard.tsx]: It's better to set `"create"` explicitly, as the switch will not compile if new types are added.

File: src/pages/project-wizard/ProjectWizard.tsx [hunk 2]
```diff
@@ -76,25 +76,37 @@ class ProjectWizardImpl extends React.Component<Props, State> {
...
+        if (this.props.action.type === "create") {
+            return getDevProject(
+                Project.create(this.props.api, this.props.config),
+                this.props.isDev
+            );
+        } else if (this.props.action.type === "edit") {
+            return Project.get(this.props.api, this.props.config, this.props.action.id);
+        } else if (this.props.action.type === "clone") {
+            return Project.clone(this.props.api, this.props.config, this.props.action.id);
+        }
```
Comment (tokland) [inline src/pages/project-wizard/ProjectWizard.tsx:99]: switch is more declarative and TS-friendly for these conditionals.

File: src/models/Project.ts [hunk 3]
```diff
@@ -666,6 +667,28 @@ class Project {
             .compact()
             .join("-");
     }
+
+    static async clone(api: D2Api, config: Config, id: Id): Promise<Project> {
+        const projectToClone = await Project.get(api, config, id);
+        const clonedDocuments = await promiseMap(projectToClone.documents, async document => {
```
Comment (tokland) [inline src/models/Project.ts:673]: Documents kept on the clone appear as REMOVED+ADD. The user may be a little confused, but probably it's not worth doing anything.

**Review by tokland — APPROVED** [review]
[code+feature review]

It works, I did some minor change. All good by me.

File: src/models/Project.ts (hunk 3 above)
Comment (Ramon-Jimenez) [reply src/models/Project.ts:673]: First time validating a development for this project, but I was confused too about this. If it is an easy fix, @eperedo, I would go ahead with it

**Review by Ramon-Jimenez — CHANGES_REQUESTED** [review]
@eperedo All good, just small changes for docs REMOVED+ADD that should be fixed if it is not too much trouble (it doesn't impact the clone action, but it looks 'odd' to the user)

**Comment by eperedo** [comment]
> @eperedo All good, just small change for docs REMOVED+ADD that should be fixed if it is not too much trouble (it doesn't impact the clone action, but it looks 'odd' to the user)

Fixed. please try again and let me know if everything works fine now.

**Review by Ramon-Jimenez — APPROVED** [review]

**Pull request merged**